### ISSUE TITLE
[bug] CSS selector issue `-moz-drag-over` nonstandard Pseudo-classes

### ISSUE THREAD

**Issue by stoivo** [sparklemotion/nokogiri#3193]
I am getting warings when I use roadie when I want inline styles. It loops over the stylessheet and inlines the styles for each selector. It also inlines the styles from normalize.css

```css
button:-moz-focusring,
[type="button"]:-moz-focusring,
[type="reset"]:-moz-focusring,
[type="submit"]:-moz-focusring {
  outline: 1px dotted ButtonText;
}
```

I created a little sample script to demonstrate the issue.
```ruby
require 'nokogiri'
doc = Nokogiri.XML('<foo></foo>')

p doc.send(:css_rules_to_xpath, ['a:hover'], {})
p doc.css('a:hover')

p doc.send(:css_rules_to_xpath, ['div:empty'], {})
p doc.css('div:empty')

p doc.send(:css_rules_to_xpath, ['a:a-moz-drag-over'], {})
p doc.css('a:a-moz-drag-over')

p doc.send(:css_rules_to_xpath, ['a:-moz-drag-over'], {})
p doc.css('a:-moz-drag-over')
```

<details>
<summary>output</summary

```
["//a[nokogiri:hover(.)]"]
[]
["//div[not(node())]"]
[]
["//a[nokogiri:a-moz-drag-over(.)]"]
[]
["//a[nokogiri:-moz-drag-over(.)]"]
/Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/nokogiri-1.16.5-arm64-darwin/lib/nokogiri/xml/searchable.rb:238:in `evaluate': ERROR: Invalid expression: //a[nokogiri:-moz-drag-over(.)] (Nokogiri::XML::XPath::SyntaxError)
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/nokogiri-1.16.5-arm64-darwin/lib/nokogiri/xml/searchable.rb:238:in `xpath_impl'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/nokogiri-1.16.5-arm64-darwin/lib/nokogiri/xml/searchable.rb:219:in `xpath_internal'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/nokogiri-1.16.5-arm64-darwin/lib/nokogiri/xml/searchable.rb:211:in `css_internal'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/nokogiri-1.16.5-arm64-darwin/lib/nokogiri/xml/searchable.rb:132:in `css'
  from /Users/simon/dev/work/dodo/fakturabank/repo/bb.rb:28:in `<main>'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/railties-7.0.8/lib/rails/commands/runner/runner_command.rb:43:in `load'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/railties-7.0.8/lib/rails/commands/runner/runner_command.rb:43:in `perform'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/thor-1.2.2/lib/thor/command.rb:27:in `run'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/thor-1.2.2/lib/thor/invocation.rb:127:in `invoke_command'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/thor-1.2.2/lib/thor.rb:392:in `dispatch'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/railties-7.0.8/lib/rails/command/base.rb:87:in `perform'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/railties-7.0.8/lib/rails/command.rb:48:in `invoke'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/railties-7.0.8/lib/rails/commands.rb:18:in `<main>'
  from <internal:/Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
  from <internal:/Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
  from /Users/simon/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
  from bin/rails:4:in `<main>'
```
</details>

As you can see it don't crash for a-moz-drag-over which is an unknown  pseudo class. Maybe it crashed because it starts on a `-`. Is this something we can fix or should I try to fix it on the outside?

**Comment by flavorjones** [comment]
@stoivo Thank you for opening this issue!

This error is coming from libxml2 (and not from Nokogiri), but let me investigate and see if we can work around it.

**Comment by flavorjones** [comment]
@stoivo Can you tell me more about your use case?

The `:-moz-drag-over` pseudoclass is not something that can be implemented without a browser, and so will never match anything. Why are you (or why is roadie) doing a search on the document that you know ahead of time will not return any matches?

I just want to make sure I understand the use case before getting too far into solving this.

**Comment by stoivo** [comment]
yeah, roadie help to inline styles for mails, since that recommended way to using css in mails. 

We take application.css and inlines those rules. It's done by parsing the css stylesheet and looping over css rule by css rule and searching for the element by css. in our application.css we have included normalize.css which has `-moz-focusring` rules. We could solve this in roadi too by not searching if the rule ends with -moz-focusring

**Comment by flavorjones** [comment]
@stoivo Thanks for giving additional context.

I read through the relevant portions of the XPath query spec. For context, pseudo-classes are implemented as function calls in XPath, and libxml2 has implemented the XPath's spec for parsing function names properly.

Unfortunately, in XPath, function names are required to start with a [`NameStartChar`](https://www.w3.org/TR/REC-xml/#NT-NameStartChar) which does not include the `-` character. There's nothing we can easily do to change that, and I don't think it's beneficial enough to warrant the work at this time.

For now, your best bet is to rescue `Nokogiri::XML::XPath::SyntaxError` and skip to the next search. Sorry I couldn't be more helpful!

**Comment by flavorjones** [comment]
Hmm, actually, please tell me if there's a way Nokogiri can signal this type of error to users earlier that would be helpful for you?

It might make sense to have the XPath translation raise an error (and not the search). Would that be helpful?

**Comment by flavorjones** [comment]
@stoivo I'm curious how the branch at https://github.com/sparklemotion/nokogiri/pull/3197 works for you?

**Comment by flavorjones** [comment]
I've created a new issue for the multiple-selector issue you described: https://github.com/sparklemotion/nokogiri/issues/3207

**Comment by stoivo** [comment]
I wasn't sure if this issue was related or not. The issue you created explains the issue very well.

I will test your branch now

**Comment by stoivo** [comment]
I tested you branch I think it's an improvement. 

I also wonder if `Nokogiri::CSS::SyntaxError` is missleading since the css selector is valid. The error is that this css selector can't be converted into an XPath selector. Would it be possible to have separate exception for that without too much work?

**Comment by flavorjones** [comment]
@stoivo I don't have any objections to creating a new subclass of Nokogiri::CSS:SyntaxError for this. Are you up for creating a pull request for it?

**Comment by stoivo** [comment]
Sounds like a fun challenge but I would need some guidance on where to start. Also I  have never worked with racc before but I can learn it.

**Comment by flavorjones** [comment]
@stoivo No need to learn rexical or racc ... it's just Ruby code. The `on_error` handler for the parser is here: https://github.com/sparklemotion/nokogiri/blob/main/lib/nokogiri/css/parser_extras.rb#L84-L87

**Comment by flavorjones** [comment]
note I reverted #3197 and instead implemented #3223

**Comment by stoivo** [comment]
The new code looks nice
